### PR TITLE
GSW-1458 feat: governance

### PR DESCRIPTION
**Pull request by r3v4s**
## feat
1. xGNS 
- most likey grc20 token (but can not be transfered)
- convert GNS to xGNS with ratio 1:1
3. Governance Staker
- stake xGNS 
- delegate / undelegate
- collect reward (gns emission, protocol_fee distribution)
- multi token support
4. Governance
- make / vote proposal
- execute other contract's function
- delegate cumulative
- emergency halt

### CODE REVIEW

File: gov/staker/staker.gno [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+type Staker struct {
+	token             grc20.Token
+	delegations       map[std.Address]*u256.Uint
+	delegationHistory map[std.Address][]DelegationHistory
+	totalStaked       *u256.Uint
+}
+
+
+
+
```
Comment (notJoon) [inline gov/staker/staker.gno:149]:
```suggestion
```

**Comment by r3v4s** [comment]
@notJoon It's draft PR. Please do not commit anything.

**Review by notJoon — DISMISSED** [review]

File: protocol_fee/protocol_fee.gno [hunk 2]
```diff
@@ -3,21 +3,30 @@ package protocol_fee
...
 
+	"gno.land/r/gnoswap/v2/common"
 	"gno.land/r/gnoswap/v2/consts"
 
 	"gno.land/p/demo/ufmt"
 )
 
-var xGnsPctBps uint64 = 5000 // 50%
+var (
+	devOpsPct      uint64 = 3000 // 30%
```
Comment (dongwon8247) [inline protocol_fee/protocol_fee.gno:13]:
```suggestion
	devOpsPct      uint64 = 0 // 0%
```

File: protocol_fee/protocol_fee.gno [hunk 3]
```diff
@@ -26,41 +35,67 @@ func DistributeProtocolFee() {
 	tokens := GetRegisteredTokens()
 	for _, token := range tokens {
 		// default distribute protocol fee percent
-		// xGNS 50%
-		// devOps 50%
+		// devOps 30%
+		// govStaker 70%
```
Comment (dongwon8247) [inline protocol_fee/protocol_fee.gno:39]:
```suggestion
		// govStaker 100%
```

File: protocol_fee/protocol_fee.gno [hunk 4]
```diff
@@ -26,41 +35,67 @@ func DistributeProtocolFee() {
...
 		// ...
 
 		balance := balanceOfByRegisterCall(token, consts.PROTOCOL_FEE_ADDR)
 		if balance > 0 {
-			toDevOps := balance * 5000 / 10000
-			toXGns := balance - toDevOps
+			toDevOps := balance * devOpsPct / 10000 // default 30%
+			toGovStaker := balance - toDevOps       // default 70%
```
Comment (dongwon8247) [inline protocol_fee/protocol_fee.gno:45]:
```suggestion
			toDevOps := balance * devOpsPct / 10000 // default 0%
			toGovStaker := balance - toDevOps       // default 100%
```

Comment (dongwon8247) [reply protocol_fee/protocol_fee.gno:45]: Would any parts of the codebase be affected by this change?

**Review by dongwon8247 — DISMISSED** [review]
Please check my comments before merging this out

File: gov/governance/config.gno [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+func init() {
+	// value for testing case from @notJoon
+	config = Config{
+		VotingStartDelay:              uint64(100),
+		VotingPeriod:                  uint64(1000),
+		VotingWeightSmoothingDuration: uint64(100),
+		Quorum:                        uint64(500),
+		ProposalCreationThreshold:     uint64(100),
```
Comment (dongwon8247) [inline gov/governance/config.gno:24]: What does this do? can we set a minimum holding of GNS of a user to create a proposal with this?

File: protocol_fee/protocol_fee.gno (hunk 4 above)
Comment (r3v4s) [reply protocol_fee/protocol_fee.gno]: test cases will fail, I'll update related ones

File: gov/governance/config.gno (hunk 5 above)
Comment (r3v4s) [reply gov/governance/config.gno:24]: purpose of each configuration value are described here
https://docs.gnoswap.io/core-concepts/governance#scope-of-proposals-5

`ProposalCreationThreshold` <- this is required amount of token for user to create proposal.

**Review by notJoon — DISMISSED** [review]
It works well 👍

**Comment by r3v4s** [comment]
changing to draft
WIP // 
1. Getter
2. Partial Undelegation History

**Review by notJoon — APPROVED** [review]

**Review by notJoon — APPROVED** [review]

**Review by onlyhyde — APPROVED** [review]
LGTM

**Pull request merged**